tests/unit/testsuite/Bolt/Boltpay/Controller/Traits/WebHookTraitTest.php

```diff
@@ -0,0 +1,441 @@
+            ->setMethods(array('isAjax', 'getParam'))
+            ->getMock();
+
+        $this->layout = $this->getMockBuilder('Mage_Core_Model_Layout')
+            ->setMethods(array('setDirectOutput'))
+            ->getMock();
+
+        $this->response = $this->getMockBuilder('Mage_Core_Controller_Response_Http')
+            ->setMethods(
+                array(
+                    'setHeader',
+                    'setBody',
+                    'setException',
+                    'clearAllHeaders',
+                    'clearBody',
+                    'setHttpResponseCode',
+                    'sendHeaders',
+                    'sendResponse'
+                )
+            )
+            ->getMock();
+
+        $this->helperMock = $this->getMockBuilder('Bolt_Boltpay_Helper_Data')
+            ->setMethods(
+                array('setResponseContextHeaders', 'verify_hook', 'notifyException')
+            )
+            ->getMock();
+
+        Mage::register('_helper/boltpay', $this->helperMock);
+
+        Mage::getModel('boltpay/observer')->initializeBenchmarkProfiler();
+
+        $this->currentMock->method('getLayout')->willReturn($this->layout);
+        $this->currentMock->method('getRequest')->willReturn($this->request);
+        $this->currentMock->method('getResponse')->willReturn($this->response);
+    }
+
+    /**
+     * Cleanup changes made by tests
+     */
+    protected function tearDown()
+    {
+        Mage::unregister('_helper/boltpay');
+        self::$_fastcgiFinishRequestCalled = false;
+        unset($_SERVER['HTTP_X_BOLT_HMAC_SHA256']);
+        Bolt_Boltpay_StreamHelper::restore();
+    }
+
+    /**
+     * @test
+     * Pre-dispatch method without validating signature
+     *
+     * @covers Bolt_Boltpay_Controller_Traits_WebHookTrait::preDispatch
+     */
+    public function preDispatch_withRequestSignatureValidationDisabled_shouldNotTryToVerifyHook()
+    {
+        $this->response->expects($this->once())->method('clearAllHeaders')->willReturnSelf();
+        $this->response->expects($this->once())->method('clearBody')->willReturnSelf();
+        $this->helperMock->expects($this->once())->method('setResponseContextHeaders');
+        $this->response->expects($this->once())->method('setHeader')->with('Content-type', 'application/json', true);
+        $this->layout->expects($this->once())->method('setDirectOutput')->with(true)
+            ->willThrowException(new Exception('Avoid calling parent::preDispatch'));
+        $this->helperMock->expects($this->never())->method('verify_hook');
+        TestHelper::setNonPublicProperty(
+            $this->currentMock,
+            'requestMustBeSigned',
+            false
+        );
+
+        //this is expected because we are mocking a trait
+        $this->setExpectedException('Exception', 'Avoid calling parent::preDispatch');
+
+        // check to disable output buffering since ob_start() is called inside preDispatch
+        if (ob_get_level() > 0) {
```

Comment:
Could we do this output buffer check in a `setUp` step? Lots of repeated code here.